integrations/nifi/nifi-rest/nifi-rest-client/nifi-rest-client-api/src/main/java/com/thinkbiganalytics/nifi/rest/client/AbstractNiFiControllerServicesRestClient.java

```diff
@@ -81,6 +95,161 @@ public ControllerServiceDTO updateStateById(@Nonnull final String id, @Nonnull f
     }
 
     /**
+     * Updates a controller service
+     * This will first disable the service, stop all referencing components, enable the service, reset the state of the components back to their prior state
+     *
+     * @param controllerService the service to update with the updated properties
+     * @return the updates service
+     */
+    @Override
+    public ControllerServiceDTO updateServiceAndReferencingComponents(ControllerServiceDTO controllerService) {
+        String id = controllerService.getId();
+        //Get all references to this controller service.  this will include processors, other controller services, and reporting tasks
+        Optional<ControllerServiceReferencingComponentsEntity> references = getReferences(id);
+
+        //state of the processors prior to update
+        Map<String, String> previousProcessorState = null;
+        //revisions for the processor references
+        Map<String, RevisionDTO> processorRevisions = null;
+
+        //revisions for the controller service references
+        Map<String, RevisionDTO> controllerServiceRevisions = null;
+
+        //a map of component id to reference entity.  this will include all referencing processors, controller services, and reporting tasks
+        Map<String, ControllerServiceReferencingComponentEntity> referencingComponentEntityMap = null;
+
+        //Stop all processor references and also disable any other controller service references
+        if (references.isPresent()) {
+            //build the reference state and revision maps prior to making this update
+            referencingComponentEntityMap = getReferencingComponents(references.get().getControllerServiceReferencingComponents());
+
+            previousProcessorState =
+                referencingComponentEntityMap.values().stream().filter(e -> e.getComponent().getReferenceType().equalsIgnoreCase("PROCESSOR")).map(c -> c.getComponent()).collect(Collectors.toMap(
+                    ControllerServiceReferencingComponentDTO::getId, ControllerServiceReferencingComponentDTO::getState));
+
+            processorRevisions = referencingComponentEntityMap.values().stream().filter(e -> e.getComponent().getReferenceType().equalsIgnoreCase("PROCESSOR")).collect(Collectors.toMap(
+                ControllerServiceReferencingComponentEntity::getId, ControllerServiceReferencingComponentEntity::getRevision));
+
+            controllerServiceRevisions =
+                referencingComponentEntityMap.values().stream().filter(e -> e.getComponent().getReferenceType().equalsIgnoreCase("ControllerService")).collect(Collectors.toMap(
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "ControllerService" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)